### PR TITLE
Example of the Doctrine doc with types as the first-class citizen

### PR DESCRIPTION
**Pull request by weaverryan**
This updates the doctrine chapter in the new "types-first" philosophy - i.e. using type-hinting and autowiring to fetch services (rather than service ids). Of course, we must still show both (the type hint and the service id), but probably only the first time in a chapter. Afterwards, we should show the type.

This is the first "normal" chapter to get this update. Thoughts?

Thanks!

### CODE REVIEW

File: doctrine.rst [hunk 1]
```diff
@@ -569,26 +571,18 @@ a controller, this is pretty easy. Add the following method to the
         return new Response('Saved new product with id '.$product->getId());
     }
 
-    // you can also receive the $em as an argument
-    public function editAction(EntityManagerInterface $em)
+    // you can use the registry to fetch other em's, if you have mutiple
+    public function editAction(ManagerRegistry $registry)
```
Comment (javiereguiluz) [inline doctrine.rst]: A question newcomers (and not only them) may ask: why sometimes I need to inject an interface (`EntityManagerInterface`) and sometimes just a class (`ManagerRegistry`).

Comment (weaverryan) [inline doctrine.rst]: Bah, that's Doctrine's fault. `ManagerRegistry` IS an interface... I think they've only more recently started to use the Interface suffix. 

Btw, I'm hoping the new `debug:container --types` option will help this out - since that valid type-hints

File: doctrine.rst [hunk 2]
```diff
@@ -638,10 +632,11 @@ Fetching an object back out of the database is even easier. For example,
...
 
-    public function showAction($productId)
+    use Doctrine\ORM\EnityManagerInterface;
+
+    public function showAction($productId, EnityManagerInterface $em)
     {
-        $product = $this->getDoctrine()
-            ->getRepository('AppBundle:Product')
+        $product = $em->getRepository('AppBundle:Product')
             ->find($productId);
```
Comment (yceruto) [inline doctrine.rst]: [DX] ``$em->find(Product::class, $productId)``? at the end it does the same, so less coding :)

File: doctrine.rst (hunk 1 above)
Comment (yceruto) [inline doctrine.rst]: ``$doctrine``?

**Review by yceruto — COMMENTED** [review]
Great! :+1: Just minor (optional) dx comments :)

File: doctrine.rst [hunk 3]
```diff
@@ -638,10 +632,11 @@ Fetching an object back out of the database is even easier. For example,
...
 on its ``id`` value::
 
-    public function showAction($productId)
+    use Doctrine\ORM\EnityManagerInterface;
+
+    public function showAction($productId, EnityManagerInterface $em)
     {
-        $product = $this->getDoctrine()
-            ->getRepository('AppBundle:Product')
+        $product = $em->getRepository('AppBundle:Product')
```
Comment (weaverryan) [inline doctrine.rst]: Yes, we should start doing this :). But it's gotta be in a different PR, because it needs to be applied to 3.1 and higher

File: doctrine.rst (hunk 1 above)
Comment (weaverryan) [inline doctrine.rst]: +1

File: doctrine.rst [hunk 4]
```diff
@@ -549,17 +549,19 @@ a controller, this is pretty easy. Add the following method to the
     use AppBundle\Entity\Product;
     use Symfony\Component\HttpFoundation\Response;
     use Doctrine\ORM\EntityManagerInterface;
+    use Doctrine\Common\Persistence\ManagerRegistry;
 
     // ...
-    public function createAction()
+    public function createAction(EntityManagerInterface $em)
```
Comment (GuilhemN) [inline doctrine.rst]: missing use statement :)

**Review by GuilhemN — APPROVED** [review]
awesome :+1:

**Comment by GuilhemN** [comment]
I'm wondering, should we assert everywhere that folders registered in the standard edition do not need to be registered in the docs?

File: doctrine.rst [hunk 5]
```diff
@@ -569,26 +571,18 @@ a controller, this is pretty easy. Add the following method to the
         return new Response('Saved new product with id '.$product->getId());
     }
 
-    // you can also receive the $em as an argument
-    public function editAction(EntityManagerInterface $em)
+    // you can use the registry to fetch other em's, if you have mutiple
```
Comment (javiereguiluz) [inline doctrine.rst]:
```
you can use the registry to fetch other em's, if you have mutiple
```

-->

```
if you have multiple entity managers, use the registry to fetch them
```

File: doctrine.rst (hunk 3 above)
Comment (weaverryan) [inline doctrine.rst]: Opened #7892 to keep track of this :)

File: doctrine.rst (hunk 2 above)
Comment (weaverryan) [inline doctrine.rst]: Ah, it's cool ... but then you haven't taught the user about the repository. So, I like the slightly longer, but "simpler" approach - tell them that you need the repo 100% of the time :)

**Comment by weaverryan** [comment]
> I'm wondering, should we assert everywhere that folders registered in the standard edition do not need to be registered again in the docs?

@GuilhemN Let's have that discussion on a different PR - like #7883, or another that I'll hopefully open today (this PR just doesn't have any good examples of that as far as I can see). But I'm totally wondering about that. Unless... we can get symfony/symfony-standard#1070 merged... then it won't matter (and this is one of the reasons I want that feature!)

File: doctrine.rst [hunk 6]
```diff
@@ -569,26 +571,18 @@ a controller, this is pretty easy. Add the following method to the
...
     }
 
-    // you can also receive the $em as an argument
-    public function editAction(EntityManagerInterface $em)
+    // if you have multiple entity managers, use the registry to fetch them
+    public function editAction(ManagerRegistry $doctrine)
     {
-        // ...
+        $em = $doctrine->getManager();
+        $em2 = $doctrine->getManager('other_connection')
```
Comment (xabbuh) [inline doctrine.rst]: I am not sure if it is common to get the manager for a particular connection. Don't you rather want to get the entity manager that manages a particular entity (i.e. you will want to use `getManagerForClass()` instead)?
